cern_access/indico_cern_access/util.py

```diff
@@ -0,0 +1,294 @@
+    else:
+        registration.cern_access_request = CERNAccessRequest(request_state=state,
+                                                             reservation_code=reservation_code)
+
+
+def create_access_request_regform(regform, state):
+    """Creates CERN access request object for registration form"""
+    from indico_cern_access.plugin import CERNAccessPlugin
+    access_tpl = CERNAccessPlugin.settings.get('access_ticket_template_id')
+    if state == CERNAccessRequestState.accepted and access_tpl:
+        regform.ticket_template = access_tpl
+    if regform.cern_access_request:
+        regform.cern_access_request.request_state = state
+    else:
+        regform.cern_access_request = CERNAccessRequestRegForm(request_state=state)
+
+
+def is_authorized_user(user):
+    """Checks if user is authorized to request access to CERN"""
+    from indico_cern_access.plugin import CERNAccessPlugin
+    return CERNAccessPlugin.settings.acls.contains_user('authorized_users', user)
+
+
+def notify_access_withdrawn(registrations):
+    """Notifies participants when access to CERN has been withdrawn"""
+    for registration in registrations:
+        template = get_template_module('cern_access:request_withdrawn_email.html', registration=registration)
+        from_address = registration.registration_form.sender_address
+        email = make_email(to_list=registration.email, from_address=from_address,
+                           template=template, html=True)
+        send_email(email, event=registration.registration_form.event, module='Registration', user=session.user)
+
+
+def send_tickets(registrations):
+    """Sends tickets to access CERN site to registered users"""
+    for registration in registrations:
+        template = get_template_module('cern_access:ticket_email.html', registration=registration)
+        from_address = registration.registration_form.sender_address
+        attachments = get_ticket_attachments(registration)
+        email = make_email(to_list=registration.email, from_address=from_address,
+                           template=template, html=True, attachments=attachments)
+        send_email(email, event=registration.registration_form.event, module='Registration',
+                   user=session.user)
+
+
+def enable_ticketing(regform):
+    """Enables ticketing module automatically"""
+    if not regform.tickets_enabled:
+        regform.tickets_enabled = True
+        regform.tickets_on_email = True
+        regform.ticket_on_event_page = True
+        regform.ticket_on_summary_page = True
+
+
+def is_category_blacklisted(category):
+    from indico_cern_access.plugin import CERNAccessPlugin
+    return any(category.id == int(cat['id']) for cat in CERNAccessPlugin.settings.get('excluded_categories'))
+
+
+def grant_access(registrations, regform):
+    event = regform.event
+    new_registrations = [reg for reg in registrations if
+                         not (reg.cern_access_request
```

Comment:
I'd move the `if` into this line, ie like this:

```python
    new_registrations = [reg for reg in registrations
                         if not (reg.cern_access_request and
                                 reg.cern_access_request.is_active and
                                 reg.cern_access_request.request_state == CERNAccessRequestState.accepted)]
```

Pretty much a matter of preference though.